pkg/query-service/ee/usage/manager.go

```diff
@@ -0,0 +1,315 @@
+	`
+	err := lm.clickhouseConn.Select(ctx, &tableSizes, query)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, val := range tableSizes {
+		switch val.Table {
+		case "logs":
+			if val.DiskName == "default" {
+				snap.CurrentLogSizeBytes = val.UncompressedBytes
+			} else {
+				snap.CurrentLogSizeBytesColdStorage = val.UncompressedBytes
+			}
+		case "samples_v2":
+			if val.DiskName == "default" {
+				snap.CurrentSamplesCount = val.Rows
+			} else {
+				snap.CurrentSamplesCountColdStorage = val.Rows
+			}
+		case "signoz_index_v2":
+			if val.DiskName == "default" {
+				snap.CurrentSpansCount = val.Rows
+			} else {
+				snap.CurrentSpansCountColdStorage = val.Rows
+			}
+		}
+	}
+
+	return &snap, nil
+}
+
+func (lm *Manager) UploadUsage(ctx context.Context) error {
+	snapshots, err := lm.repository.GetSnapshotsNotSynced(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(snapshots) <= 0 {
+		zap.S().Info("no snapshots to upload, skipping.")
+		return nil
+	}
+
+	zap.S().Info("uploading snapshots")
+	for _, snap := range snapshots {
+		metricsBytes, err := encryption.Decrypt([]byte(snap.ActivationId.String()[:32]), []byte(snap.Snapshot))
+		if err != nil {
+			return err
+		}
+
+		metrics := model.UsageSnapshot{}
+		err = json.Unmarshal(metricsBytes, &metrics)
+		if err != nil {
+			return err
+		}
+
+		err = lm.UploadUsageWithExponentalBackOff(ctx, model.UsagePayload{
+			UsageBase: model.UsageBase{
+				Id:                snap.Id,
+				InstallationId:    snap.InstallationId,
+				ActivationId:      snap.ActivationId,
+				FailedSyncRequest: snap.FailedSyncRequest,
+			},
+			SnapshotDate: snap.CreatedAt,
+			Metrics:      metrics,
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (lm *Manager) UploadUsageWithExponentalBackOff(ctx context.Context, payload model.UsagePayload) error {
+	for i := 1; i <= MaxRetries; i++ {
+		apiErr := licenseserver.SendUsage(ctx, &payload)
+		if apiErr != nil && i == MaxRetries {
+			err := lm.repository.IncrementFailedRequestCount(ctx, payload.Id)
+			if err != nil {
+				zap.S().Errorf("failed to updated the failure count for snapshot in DB : ", zap.Error(err))
+				return err
+			}
+			zap.S().Errorf("retries stopped : %v", zap.Error(err))
+			// not returning error here since it is captured in the failed count
+			return nil
+		} else if apiErr != nil {
+			// sleeping for exponential backoff
+			sleepDuration := RetryInterval * time.Duration(i)
+			zap.S().Errorf("failed to upload snapshot retrying after %v secs : %v", sleepDuration.Seconds(), zap.Error(apiErr.Err))
+			time.Sleep(sleepDuration)
+
+			// update the failed request count
+			err := lm.repository.IncrementFailedRequestCount(ctx, payload.Id)
+			if err != nil {
+				zap.S().Errorf("failed to updated the failure count for snapshot in DB : %v", zap.Error(err))
+				return err
+			}
+		} else {
+			break
+		}
+	}
+
+	// update the database that it is synced
+	err := lm.repository.MoveToSynced(ctx, payload.Id)
```

Comment:
I think you should make this a transaction with first update this to synced and then roll back if the upload is not successful in `MaxRetries` times.

Reply:
Don't think a transaction is required as this is a single update query.